server/src/main/java/com/tencent/rss/server/ShuffleServerGrpcService.java

```diff
@@ -426,21 +428,63 @@ public void getShuffleData(GetShuffleDataRequest request,
     responseObserver.onCompleted();
   }
 
-  private List<ShuffleDataBlockSegment> toBlockSegments(List<BufferSegment> bufferSegments) {
-    List<ShuffleDataBlockSegment> ret = Lists.newArrayList();
-
-    for (BufferSegment segment : bufferSegments) {
-      ret.add(ShuffleDataBlockSegment.newBuilder()
-          .setBlockId(segment.getBlockId())
-          .setOffset(segment.getOffset())
-          .setLength(segment.getLength())
-          .setUncompressLength(segment.getUncompressLength())
-          .setCrc(segment.getCrc())
-          .setTaskAttemptId(segment.getTaskAttemptId())
-          .build());
+  @Override
+  public void getShuffleIndex(GetShuffleIndexRequest request,
+      StreamObserver<GetShuffleIndexResponse> responseObserver) {
+    ShuffleServerMetrics.counterTotalRequest.inc();
+
+    String appId = request.getAppId();
+    int shuffleId = request.getShuffleId();
+    int partitionId = request.getPartitionId();
+    int partitionNumPerRange = request.getPartitionNumPerRange();
+    int partitionNum = request.getPartitionNum();
+    StatusCode status = StatusCode.SUCCESS;
+    String msg = "SUCCESS";
+    GetShuffleIndexResponse reply;
+    String requestInfo = "appId[" + appId + "], shuffleId[" + shuffleId + "], partitionId["
+        + partitionId + "]";
+
+    if (shuffleServer.getMultiStorageManager() != null) {
+      shuffleServer.getMultiStorageManager().updateLastReadTs(appId, shuffleId, partitionId);
     }
 
-    return ret;
+    if (shuffleServer.getShuffleBufferManager().requireReadMemoryWithRetry(request.getLength())) {
+      try {
+        long start = System.currentTimeMillis();
+        ShuffleIndexResult shuffleIndexResult = shuffleServer.getShuffleTaskManager().getShuffleIndex(
+            appId, shuffleId, partitionId, partitionNumPerRange, partitionNum);
+        long readTime = System.currentTimeMillis() - start;
+
+        ShuffleServerMetrics.counterTotalReadTime.inc(readTime);
+        GetShuffleIndexResponse.Builder builder = GetShuffleIndexResponse.newBuilder()
+            .setStatus(valueOf(status))
+            .setRetMsg(msg);
+        if (!shuffleIndexResult.isEmpty()) {
+          builder.setIndexData(ByteString.copyFrom(shuffleIndexResult.getIndexData()));
```

Comment:
builder.setIndexData should be if there is no data, or exception happened

Reply:
OK